lib/auth.ts

```diff
@@ -133,6 +152,82 @@ export class Auth {
     return token;
   };
 
+  /**
+   * Perform a request against Kontist REST API
+   */
+  private request = async (path: string, method: HttpMethod, body?: string) => {
+    if (!this.token) {
+      throw new UserUnauthorizedError();
+    }
+
+    const requestUrl = new URL(path, this.baseUrl).href;
+
+    const response = await fetch(requestUrl, {
+      method,
+      headers: {
+        Accept: "application/json",
+        "Content-Type": "application/json",
+        Authorization: `Bearer ${this.token.accessToken}`
+      },
+      body
+    });
+
+    if (!response.ok) {
+      throw new KontistSDKError({
+        status: response.status,
+        message: response.statusText
+      })
+    }
+
+    return response.json();
+  };
+
+  /**
+   * Called periodically by `getMFAConfirmedToken` until
+   * the challenge expires or its status is updated
+   */
+  private pollChallengeStatus = (
+    pendingChallenge: Challenge,
+    resolve: Function,
+    reject: Function
+  ) => async () => {
+    const challenge = await this.request(
+      `${MFA_CHALLENGE_PATH}/${pendingChallenge.id}`,
+      HttpMethod.GET
+    );
+
+    if (new Date(challenge.expiresAt) < new Date()) {
+      clearInterval(this.challengePollIntervalId as IntervalID);
+      return reject(new ChallengeExpiredError());
+    } else if (challenge.status === ChallengeStatus.DENIED) {
+      clearInterval(this.challengePollIntervalId as IntervalID);
+      return reject(new ChallengeDeniedError());
+    } else if (challenge.status === ChallengeStatus.VERIFIED) {
+      clearInterval(this.challengePollIntervalId as IntervalID);
+      const { token: confirmedToken } = await this.request(
+        `${MFA_CHALLENGE_PATH}/${challenge.id}/token`,
+        HttpMethod.POST
+      );
+
+      const token = this.setToken(confirmedToken);
+      return resolve(token);
+    }
+  };
+
+  /**
+   * Create an MFA challenge and request a confirmed access token when verified
+   */
+  public getMFAConfirmedToken = async () => {
+    const challenge = await this.request(MFA_CHALLENGE_PATH, HttpMethod.POST);
+
+    return new Promise((resolve, reject) => {
+      this.challengePollIntervalId = setInterval(
+        this.pollChallengeStatus(challenge, resolve, reject),
+        this.challengePollInterval
+      );
```

Comment:
This polling implementation can have multiple parallel requests if user has very slow internet connection or `challengePollInterval` is less than request time. Is it ok for us?